docs/adr/0004-workflow-instance-indexing-and-trigger-acceptance.md

```diff
@@ -0,0 +1,61 @@
+# Workflow Instance Indexing and Trigger Acceptance
+
+## Status
+
+Proposed
+
+## Context
+
+### Background - current method for matching and processing triggers
+
+As described in ADRs 0001-0003, we are rewriting a section of the workflow engine using the event sourcing pattern. This forces us to consider some aspects of the way in which triggers are processed, especially when workflow instances raise triggers targetting themselves.
+
+At present, the process for handling a trigger is as follows:
+- Search for workflow instances whose `Interests` collection has at least one element in common with the trigger `Subjects`.
+- For each of those instances
+  - Acquire a lease using the workflow instance Id to prevent any other triggers for the same instance from being processed concurrently
+  - Process the trigger
+  - Persist the updated workflow instance
+  - Release the lease
+
+The act of processing a trigger can cause further triggers to be raised. These are raised through the message ingestion endpoint as normal. If a trigger raised in this way targets the workflow instance that caused it to be raised, the leasing process prevents it for being processed until the original trigger processing is complete.
+
+In this specific scenario, it is highly likely that triggers raised in this way will be matched against workflow instance interests that are not up to date, because (assuming they are recieved and processed prior to processing of the original trigger) they will be matched against interests as they were prior to the original trigger being executed. Historically we've been able to ignore this issue because the workflow instance Id is always part of a workflow's `Interests`.
```

Comment:
It is undesirable to couple Application triggers to specific workflow (instances) in this way, as the Trigger lives in the Application domain, and the Workflow definition (or definitions) may respond to that trigger in any number of different ways over the lifetime of the application. That is part of the purpose of separating the workflow engine out like this.

The best example of this is perhaps dropping a section of a worklow into multiple parallel subworkflows with a "first wins". Those triggers that once took a single path through the outer workflow may now advance more than one workflow instance at a time.

Reply:
This paragraph's talking specifically about the case when a workflow instance is raising a trigger intended "for itself" - so I guess this doesn't qualify as an "application trigger". However. I agree with your point. I think we can encourage this separation by disallowing instance creation requests from specifying a workflow instance Id - but we can't do this properly until we have added the ability to add extra data to a workflow instance's interests collection. Once this is done, we can add relevant application-specific entity Ids to the interests collection allowing us to target triggers at those entities rather than specific workflow instances.